### PR TITLE
Enable pass query string to input_map in ml inference search response processor

### PR DESCRIPTION
**Pull request by mingshl**
### Description
Enable pass query string to input_map in ml inference search response processor

## setting cluster 
```
PUT _cluster/settings
{
  "persistent": {
    "plugins": {
      "ml_commons": {
        "only_run_on_ml_node": "false",
        "model_access_control_enabled": "true",
        "native_memory_threshold": "99"
      }
    }
  }
}
```

## register cross-encoders local model 
```
POST /_plugins/_ml/models/_register
{
  "name": "huggingface/cross-encoders/ms-marco-MiniLM-L-6-v2",
  "version": "1.0.2",
  "model_format": "TORCH_SCRIPT"
}

```

```
{
  "task_id": "tQ5p1ZEB4iWlnHsIf2Xw",
  "status": "CREATED"
}

```

## get register task status 
`GET /_plugins/_ml/tasks/tQ5p1ZEB4iWlnHsIf2Xw`

```
{
  "model_id": "tg5p1ZEB4iWlnHsIh2U9",
  "task_type": "REGISTER_MODEL",
  "function_name": "TEXT_SIMILARITY",
  "state": "COMPLETED",
  "worker_node": [
    "AahcbpI9R2OtId7Wnt0cYA"
  ],
  "create_time": 1725862346467,
  "last_update_time": 1725862356009,
  "is_async": true
}
```

## deploy cross-encoders local model 
`POST /_plugins/_ml/models/tg5p1ZEB4iWlnHsIh2U9/_deploy

`
```
{
  "task_id": "tw5q1ZEB4iWlnHsIo2WX",
  "task_type": "DEPLOY_MODEL",
  "status": "CREATED"
}
```

## get deploy task status
`GET /_plugins/_ml/tasks/tw5q1ZEB4iWlnHsIo2WX`

```
 {
  "model_id": "tg5p1ZEB4iWlnHsIh2U9",
  "task_type": "DEPLOY_MODEL",
  "function_name": "TEXT_SIMILARITY",
  "state": "RUNNING",
  "worker_node": [
    "AahcbpI9R2OtId7Wnt0cYA"
  ],
  "create_time": 1725862421392,
  "last_update_time": 1725862421555,
  "is_async": true
}
```

## wait until completed
```
{
  "model_id": "tg5p1ZEB4iWlnHsIh2U9",
  "task_type": "DEPLOY_MODEL",
  "function_name": "TEXT_SIMILARITY",
  "state": "COMPLETED",
  "worker_node": [
    "AahcbpI9R2OtId7Wnt0cYA"
  ],
  "create_time": 1725862421392,
  "last_update_time": 1725862466568,
  "is_async": true
}
```

  
## test model predict
```
POST _plugins/_ml/models/tg5p1ZEB4iWlnHsIh2U9/_predict
{
    "query_text": "today is sunny",
    "text_docs": [
        "how are you",
        "today is sunny",
        "today is july fifth",
        "it is winter"
    ]
}

```
```
{
  "inference_results": [
    {
      "output": [
        {
          "name": "similarity",
          "data_type": "FLOAT32",
          "shape": [
            1
          ],
          "data": [
            -11.055183
          ],
          "byte_buffer": {
            "array": "COIwwQ==",
            "order": "LITTLE_ENDIAN"
          }
        }
      ]
    },
    {
      "output": [
        {
          "name": "similarity",
          "data_type": "FLOAT32",
          "shape": [
            1
          ],
          "data": [
            8.969885
          ],
          "byte_buffer": {
            "array": "poQPQQ==",
            "order": "LITTLE_ENDIAN"
          }
        }
      ]
    },
    {
      "output": [
        {
          "name": "similarity",
          "data_type": "FLOAT32",
          "shape": [
            1
          ],
          "data": [
            -5.736347
          ],
          "byte_buffer": {
            "array": "KJC3wA==",
            "order": "LITTLE_ENDIAN"
          }
        }
      ]
    },
    {
      "output": [
        {
          "name": "similarity",
          "data_type": "FLOAT32",
          "shape": [
            1
          ],
          "data": [
            -10.0452175
          ],
          "byte_buffer": {
            "array": "NrkgwQ==",
            "order": "LITTLE_ENDIAN"
          }
        }
      ]
    }
  ]
}
```
## upload index
```
 PUT /demo-index-0/_doc/1
{
  "dairy": "how are you"
}

```
```
 PUT /demo-index-0/_doc/2
{
  "dairy": "today is sunny"
}
```

```
 PUT /demo-index-0/_doc/3
{
  "dairy": "today is july fifth"
}
```

```
{
  "_index": "demo-index-0",
  "_id": "4",
  "_version": 1,
  "result": "created",
  "_shards": {
    "total": 2,
    "successful": 1,
    "failed": 0
  },
  "_seq_no": 3,
  "_primary_term": 1
}
```

## create search pipeline with query text passing in input_map

```
PUT /_search/pipeline/my_pipeline
{
  "response_processors": [
    {
      "ml_inference": {
        "tag": "ml_inference",
        "description": "This processor is going to run ml inference during search response",
        "model_id": "tg5p1ZEB4iWlnHsIh2U9",
        "model_input":"{ \"text_docs\": ${input_map.text_docs}, \"query_text\": \"${input_map.query_text}\" }",
        "function_name": "TEXT_SIMILARITY",
        "input_map": [
          {
            "text_docs": "dairy",
             "query_text": "$.query.term.dairy.value"          
         }
        ],
        "output_map": [
          {
            "rank_score": "$.inference_results[*].output[*].data"
          }
        ],
        "full_response_path":false,
        "model_config": {
        },
        "ignore_missing": false,
        "ignore_failure": false
      }
    }
  ]
}

```
## search with search pipeline, scores are added in the response
```
GET /demo-index-0/_search?search_pipeline=my_pipeline
{
  "query": {
    "term": {
      "dairy": {
        "value": "today"
      }
    }
  }
}
```

```
{
  "took": 400,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 2,
      "relation": "eq"
    },
    "max_score": 0.71566814,
    "hits": [
      {
        "_index": "demo-index-0",
        "_id": "2",
        "_score": 0.71566814,
        "_source": {
          "dairy": "today is sunny",
          "rank_score": [
            3.6144485
          ]
        }
      },
      {
        "_index": "demo-index-0",
        "_id": "3",
        "_score": 0.6333549,
        "_source": {
          "dairy": "today is july fifth",
          "rank_score": [
            3.6144485
          ]
        }
      }
    ]
  }
}

```

## ToDo
currently ml inference processor only support single tensor for local model, need to support multiple tensor parsing as well. 

### Related Issues
https://github.com/opensearch-project/ml-commons/issues/2897
https://github.com/opensearch-project/ml-commons/issues/2878
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ylwu-amzn** [comment]
Add more details/examples to description ?

**Comment by ylwu-amzn** [comment]
```
> Task :opensearch-ml-common:test

StringUtilsTest > testisValidJSONPath_InvalidInputs FAILED
    java.lang.AssertionError at StringUtilsTest.java:476

StringUtilsTest > testisValidJSONPath_EmptyInput FAILED
    java.lang.AssertionError at StringUtilsTest.java:490

1008 tests completed, 2 failed, 2 skipped
```

**Comment by mingshl** [comment]
> #2897

there are more details in the RFC https://github.com/opensearch-project/ml-commons/issues/2897

**Comment by mingshl** [comment]
CI failed

```
OpenSearchConversationalMemoryHandlerTests > classMethod FAILED
Error: Exception in thread "Thread-4" java.lang.NoClassDefFoundError: Could not initialize class org.opensearch.test.OpenSearchTestCase
	at java.base/java.lang.Thread.run(Thread.java:1583)
	Suppressed: java.lang.IllegalStateException: No context information for thread: Thread[id=29, name=Thread-4, state=RUNNABLE, group=TGRP-ConversationalMemoryHandlerITTests]. Is this thread running under a class com.carrotsearch.randomizedtesting.RandomizedRunner runner context? Add @RunWith(class com.carrotsearch.randomizedtesting.RandomizedRunner.class) to your test class. Make sure your code accesses random contexts within @BeforeClass and @AfterClass boundary (for example, static test class initializers are not permitted to access random contexts).
		at com.carrotsearch.randomizedtesting.RandomizedContext.context(RandomizedContext.java:249)
		at com.carrotsearch.randomizedtesting.RandomizedContext.current(RandomizedContext.java:134)
		at com.carrotsearch.randomizedtesting.RandomizedRunner.augmentStackTrace(RandomizedRunner.java:1885)
		at com.carrotsearch.randomizedtesting.RunnerThreadGroup.uncaughtException(RunnerThreadGroup.java:20)
		at java.base/java.lang.Thread.dispatchUncaughtException(Thread.java:2901)
Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.ExceptionInInitializerError [in thread "SUITE-ConversationalMemoryHandlerITTests-seed#[7B60246E9722A278]"]
	at org.opensearch.test.OpenSearchTestCase.<clinit>(OpenSearchTestCase.java:285)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:534)
	at java.base/java.lang.Class.forName(Class.java:513)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$2.run(RandomizedRunner.java:623)
    java.lang.NoClassDefFoundError at Class.java:-2
        Caused by: java.lang.ExceptionInInitializerError at OpenSearchTestCase.java:285


```

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java [hunk 1]
```diff
@@ -293,4 +294,45 @@ public static String getJsonPath(String jsonPathWithSource) {
...
+     *     <li>"foo..bar"</li>
+     *     <li>"."</li>
+     *     <li>".."</li>
+     * </ul>
+     *
+     * @param input the input string to be checked
+     * @return true if the input string matches the dot path format, false otherwise
+     */
+    public static boolean isValidJSONPath(String input) {
+        if (input == null || input.equals("")) {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:327]: what if `" "`? may be we can use `input.isBlank()`?

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java [hunk 2]
```diff
@@ -293,4 +294,45 @@ public static String getJsonPath(String jsonPathWithSource) {
         // Extract the substring from the startIndex to the end of the input string
         return (startIndex != -1) ? jsonPathWithSource.substring(startIndex) : jsonPathWithSource;
     }
+
+    /**
+     * Checks if the given input string matches the dot path format.
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:299]: are we matching dot path format or json path format? Description and method names both are different to each other.

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java [hunk 3]
```diff
@@ -293,4 +294,45 @@ public static String getJsonPath(String jsonPathWithSource) {
...
+     *
+     * @param input the input string to be checked
+     * @return true if the input string matches the dot path format, false otherwise
+     */
+    public static boolean isValidJSONPath(String input) {
+        if (input == null || input.equals("")) {
+            return false;
+        }
+        try {
+            JsonPath.compile(input); // This will throw an exception if the path is invalid
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:331]: If we are checking only dot path format, we can easily do:

```
    // Regular expression for a valid dot path (letters, digits, underscores, separated by dots)
    String regex = "^\\.?\\w+(\\.\\w+)*\\.?$";
    return input.matches(regex);
```

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java [hunk 4]
```diff
@@ -293,4 +294,45 @@ public static String getJsonPath(String jsonPathWithSource) {
...
+     * <ul>
+     *     <li>"foo..bar"</li>
+     *     <li>"."</li>
+     *     <li>".."</li>
+     * </ul>
+     *
+     * @param input the input string to be checked
+     * @return true if the input string matches the dot path format, false otherwise
+     */
+    public static boolean isValidJSONPath(String input) {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:326]:
```
Bracket Notation (Useful for property names with special characters):

$.store['book']
$['store']['book'][0]['title']
Array Indexing:

$.store.book[0]
$.store.book[1,2] (Multiple indices)
$.store.book[-1:] (The last element)
$.store.book[0:2] (Slice array from index 0 to 2)
Wildcard (*):

$.store.book[*] (All books)
$.* (All properties at the root level)
Filters (?()):

$.store.book[?(@.price < 10)] (Books cheaper than 10)
$.store.book[?(@.author == 'J.K. Rowling')] (Books authored by J.K. Rowling)
Recursive Descent (..):

$..author (Find all authors, no matter where they are in the document)
$..book[?(@.price > 15)] (Find books with a price greater than 15 anywhere in the document)
Union ([,]):

$.store.book[0,1] (First and second books)
$['store','warehouse'] (Select multiple properties at the same level)
Combination of Wildcard and Filters:

$..book[?(@.price > 20)].title (Get the titles of books costing more than 20)
```


Are we going to validate these kinds of json path too?

OR is it just a simple dot path format? Either reflect documentation or the method name accordingly.

**Comment by dhrubo-os** [comment]
I also left a comment in your RFC.

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java (hunk 2 above)
Comment (mingshl) [reply common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:299]: right, json path is more straight forward. I will change to  json path format

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java (hunk 3 above)
Comment (mingshl) [reply common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:331]: I tried to use regex function earlier but in a deeper thought, I think we need to support all kinds of json path. Some of the json path have some nice functions too, like 'length()', we should support and confirm this is a valid json path. Checkout the function here https://github.com/json-path/JsonPath

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java (hunk 4 above)
Comment (mingshl) [reply common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:326]: we need to support all kinds of json, that's why is called `isValidJsonPath`

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java (hunk 1 above)
Comment (mingshl) [reply common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java]: good catch, changed to input.isBlank()

File: common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java [hunk 5]
```diff
@@ -457,4 +458,35 @@ public void testGetJsonPath_ValidJsonPathWithoutSource() {
         String result = getJsonPath(input);
         assertEquals("$.response.body.data[*].embedding", result);
     }
+
+    @Test
+    public void testisValidJSONPath_ValidInputs() {
```
Comment (dhrubo-os) [inline common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java:463]: As I provided more complex examples, can we add those in the test too?

Comment (mingshl) [reply common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java:463]: you mean these ones?
$.store['book']
$['store']['book'][0]['title']
Array Indexing:

$.store.book[0]
$.store.book[1,2] (Multiple indices)
$.store.book[-1:] (The last element)
$.store.book[0:2] (Slice array from index 0 to 2)
Wildcard (*):

$.store.book[*] (All books)
$.* (All properties at the root level)
Filters (?()):

$.store.book[?(@.price < 10)] (Books cheaper than 10)
$.store.book[?(@.author == 'J.K. Rowling')] (Books authored by J.K. Rowling)
Recursive Descent (..):

$..author (Find all authors, no matter where they are in the document)
$..book[?(@.price > 15)] (Find books with a price greater than 15 anywhere in the document)
Union ([,]):

$.store.book[0,1] (First and second books)
$['store','warehouse'] (Select multiple properties at the same level)
Combination of Wildcard and Filters:

$..book[?(@.price > 20)].title (Get the titles of books costing more than 20)

Comment (dhrubo-os) [reply common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java:463]: Yeah, if you think some of them could be related to your use case?

Comment (mingshl) [reply common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java:463]: sounds great, added in
[add more tests](https://github.com/opensearch-project/ml-commons/pull/2899/commits/576c3fa7f60591299aac494adc991fef84b9d20b)

**Comment by mingshl** [comment]
flaky test 

```
REPRODUCE WITH: ./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.RestMLPredictionActionTests.testGetRequest_LocalModelInferenceDisabled" -Dtests.seed=9770CFDD00A92E2A -Dtests.security.manager=false -Dtests.locale=az-AZ -Dtests.timezone=Asia/Sakhalin -Druntime.java=21

RestMLPredictionActionTests > testGetRequest_LocalModelInferenceDisabled FAILED
    java.lang.AssertionError: 
    Expected: (an instance of java.lang.IllegalStateException and exception with message a string containing "Local Model is currently disabled. To enable it, update the setting \"plugins.ml_commons.local_model.enabled\" to true.")
         but: an instance of java.lang.IllegalStateException <java.lang.IllegalArgumentException: Wrong Action Type of models> is a java.lang.IllegalArgumentException
    Stacktrace was: java.lang.IllegalArgumentException: Wrong Action Type of models
    	at org.opensearch.ml.common.connector.ConnectorAction$ActionType.from(ConnectorAction.java:199)
    	at org.opensearch.ml.rest.RestMLPredictionAction.getRequest(RestMLPredictionAction.java:129)
    	at org.opensearch.ml.rest.RestMLPredictionActionTests.testGetRequest_LocalModelInferenceDisabled(RestMLPredictionActionTests.java:146)
    	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
    	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
    	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
    	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
    	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
    	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
    	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:258)
    	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
    	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
    	at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
```

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 6]
```diff
@@ -316,22 +321,42 @@ private void rewriteResponseDocuments(SearchResponse response, ActionListener<Se
      * @param inputMapIndex           the index of the input mapping to process
      * @param batchPredictionListener the listener to be notified when the predictions are processed
      * @param hitCountInPredictions   a map to keep track of the count of hits that have the required input fields for each round of prediction
+     * @param queryString
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:324]: minor: add some explanation

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 7]
```diff
@@ -316,22 +321,42 @@ private void rewriteResponseDocuments(SearchResponse response, ActionListener<Se
...
-            modelParameters.putAll(inferenceProcessorAttributes.getModelConfigMaps());
-            modelConfigs.putAll(inferenceProcessorAttributes.getModelConfigMaps());
+            Map<String, String> modelConfigMapsInput = inferenceProcessorAttributes.getModelConfigMaps();
+
+            for (Map.Entry<String, String> entry : modelConfigMapsInput.entrySet()) {
+                String modelConfigKey = entry.getKey();
+                String modelConfigValue = entry.getValue();
+                if (StringUtils.isValidJSONPath(modelConfigValue)) {
+                    Object queryJson = JsonPath.parse(queryString).read("$");
+                    Configuration configuration = Configuration
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:347]: Not efficient to construct same Configuration in a for loop. Move this out of for loop or create a static variable ?

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 8]
```diff
@@ -316,22 +321,42 @@ private void rewriteResponseDocuments(SearchResponse response, ActionListener<Se
...
         if (inferenceProcessorAttributes.getModelConfigMaps() != null) {
-            modelParameters.putAll(inferenceProcessorAttributes.getModelConfigMaps());
-            modelConfigs.putAll(inferenceProcessorAttributes.getModelConfigMaps());
+            Map<String, String> modelConfigMapsInput = inferenceProcessorAttributes.getModelConfigMaps();
+
+            for (Map.Entry<String, String> entry : modelConfigMapsInput.entrySet()) {
+                String modelConfigKey = entry.getKey();
+                String modelConfigValue = entry.getValue();
+                if (StringUtils.isValidJSONPath(modelConfigValue)) {
+                    Object queryJson = JsonPath.parse(queryString).read("$");
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:346]: Why need to read "$" first ? , Can we just read the json path directly?
```
JsonPath.parse(queryString).read(modelConfigValue)
```

Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java]: right, addressed in commit [use standard json path config](https://github.com/opensearch-project/ml-commons/pull/2899/commits/3fd5cdb54a42148618ee1880f8677e656b21de05)

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java (hunk 6 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java]: add explanation in commit  
[add example in javadoc](https://github.com/opensearch-project/ml-commons/pull/2899/commits/98dfbec04af434462193653fe07f0ce6d8435df7)

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 9]
```diff
@@ -316,22 +321,37 @@ private void rewriteResponseDocuments(SearchResponse response, ActionListener<Se
...
-            modelParameters.putAll(inferenceProcessorAttributes.getModelConfigMaps());
-            modelConfigs.putAll(inferenceProcessorAttributes.getModelConfigMaps());
+            Map<String, String> modelConfigMapsInput = inferenceProcessorAttributes.getModelConfigMaps();
+
+            for (Map.Entry<String, String> entry : modelConfigMapsInput.entrySet()) {
+                String modelConfigKey = entry.getKey();
+                String modelConfigValue = entry.getValue();
+                if (StringUtils.isValidJSONPath(modelConfigValue)) {
+                    Object querySubString = JsonPath.using(suppressExceptionConfiguration).parse(queryString).read(modelConfigValue);
+                    if (querySubString != null) {
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:347]: In the example if model config value is valid json path like `"query_text": "$.query.term.dairy.value"`, we will always parse the value, then this line checks if the value of the query path is null or not. If the value is null, we just do nothing, so the `query_text` will still be `"$.query.term.dairy.value"`, then the model will calculate similarity score of `"$.query.term.dairy.value"` and each doc, right ? This looks somehow hide such error and not easy to figure out

Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:347]: let me add unit test for the querySubString is null when we  cannot find the query_text in the query string and test the error message

Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:347]: Is `querySubString` the value of the query path ? The name seems confusing

Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java]: renamed to `queryText`

**Comment by mingshl** [comment]
after discussion, decided to read the query string from input_map, similar to search request processor. requested review again @ylwu-amzn

File: plugin/src/main/java/org/opensearch/ml/processor/ModelExecutor.java [hunk 10]
```diff
@@ -282,8 +282,12 @@ default String toString(Object originalFieldValue) {
     }
 
     default boolean hasField(Object json, String path) {
-        Object value = JsonPath.using(suppressExceptionConfiguration).parse(json).read(path);
-
+        Object value;
+        if (json instanceof String) {
+            value = JsonPath.using(suppressExceptionConfiguration).parse((String) json).read(path);
+        } else {
+            value = JsonPath.using(suppressExceptionConfiguration).parse(json).read(path);
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/processor/ModelExecutor.java:289]: What's the difference with line 287?

Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/ModelExecutor.java:289]: if passing over a string, JsonPath requires casting to String, or it won't parse Correctly. 

String is acceptable: 

```
String jsonString = "{"text": "foo"}"
Object value = JsonPath.using(suppressExceptionConfiguration).parse(jsonString).read(path);

Object is acceptable: 
```
Map<String, String> jsonObject = new HashMap<>();
jsonObject.put("text","foo")； 
Object value = JsonPath.using(suppressExceptionConfiguration).parse(jsonObject).read(path);

```

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 11]
```diff
@@ -155,6 +154,8 @@ public void processResponseAsync(
         try {
             SearchHit[] hits = response.getHits().getHits();
             // skip processing when there is no hit
+
+            String queryString = request.source().toString();
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:158]: Does this also have `size` `sort` , or just contain `query` part ?

**Review by ylwu-amzn — APPROVED** [review]

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java (hunk 11 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:159]: added test testProcessResponseSuccessReadRequestMetaFieldFromInputMap

**Review by Zhangxunmt — APPROVED** [review]

**Comment by mingshl** [comment]
backport failed because of the previous backport 2.x is not merged yet. https://github.com/opensearch-project/ml-commons/pull/3127
